### PR TITLE
[WIP] Convert async I/O backend to Socketry

### PR DESCRIPTION
**Pull request by tarcieri**
Socketry uses async I/O to implement thread-safe timeouts:

https://github.com/socketry/socketry

cc @ixti @zanker @TiagoCardoso1983

### CODE REVIEW

File: lib/http/connection.rb [hunk 1]
```diff
@@ -37,13 +37,22 @@ def initialize(req, options)
...
+      end
+
+      @socket.connect(req.socket_host, req.socket_port)
+      @socket.nodelay = options.nodelay if options.nodelay
 
       send_proxy_connect_request(req)
-      start_tls(req, options)
       reset_timer
-    rescue IOError, SocketError, SystemCallError => ex
+    rescue Socketry::Error => ex
```
Comment (tarcieri) [inline lib/http/connection.rb]: There are probably more places this sort of change needs to be made

**Comment by tarcieri** [comment]
Note: Socketry's timeouts can subsume `HTTP::Timeout::Null` and `HTTP::Timeout::PerOperation`, but `HTTP::Timeout::Global` will need to be (re)implemented.

**Comment by tarcieri** [comment]
Note per cb420da this is a breaking change which will require a new major release (3.x) as among other things it will require Ruby 2.2+ due to the Socketry dependency.

Since Rails 5 is also 2.2+, hopefully this is ok.

File: lib/http/connection.rb [hunk 2]
```diff
@@ -142,22 +151,6 @@ def expired?
 
     private
 
-    # Sets up SSL context and starts TLS if needed.
-    # @param (see #initialize)
-    # @return [void]
-    def start_tls(req, options)
-      return unless req.uri.https? && !failed_proxy_connect?
```
Comment (zanker) [inline lib/http/connection.rb]: I suspect it is, I didn't add that at least.

**Comment by tarcieri** [comment]
Well, still a WIP (especially timeout support) but it's passing consistently on 2.3 and JRuby 9.1.6.0 (and sporadically on 2.2)

**Comment by tarcieri** [comment]
If anyone wants to resume work on this, let me know.

**Pull request closed without merging**